### PR TITLE
Make account menu configurable

### PR DESCRIPTION
**Pull request by xmorave2**
The purpose of this PR is to make a menu account a bit more flexible in custom instances of VuFind. 

I tried to make most of menu items features configurable, but there are some conditions which cannot be set easily through configuration, so they are still coded in the templates.

### CODE REVIEW

**Comment by ThoWagen** [comment]
I am currently working on something similar. I plan to add configurable links to the menu and other places that e.g. can also be links to other websites or make it possible to open a link in the lightbox. 

Please let me know if you are planing something similar as a followup PR. 

In any case I will wait with creating an own PR until this one is completed.

File: config/vufind/MyResearchMenu.ini
Comment (demiankatz) [inline config/vufind/MyResearchMenu.ini]: I think two sentences might need to be joined together here. Should it read:

"If there are no items defined, or if this file is missing, VuFind will use default settings..."

File: config/vufind/MyResearchMenu.ini
Comment (demiankatz) [inline config/vufind/MyResearchMenu.ini]: Might be worth noting required fields and default values here -- e.g. I assume that label and route are required, and status defaults to false.

File: themes/bootstrap3/templates/myresearch/menu.phtml
Comment (demiankatz) [inline themes/bootstrap3/templates/myresearch/menu.phtml]: It might make sense to add a myResearchMenu view helper. This would allow you to encapsulate the default configuration outside of the template, and it has other advantages too -- for example, you could have a config setting that could be used to call arbitrary methods of the helper to check for conditional display, and that might make it possible to encapsulate more of the special-case logic. Users could also then extend the helper to define their own rules if they want to.

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks, @xmorave2 -- I haven't studied all the code changes in close detail yet; I just took a first pass to read the documentation and get a general idea of your approach. I've left some comments suggesting minor documentation improvements and a possibly helpful idea for refactoring the rest. Please let me know what you think!

**Comment by demiankatz** [comment]
One other thought: I don't really like the "MyResearch" terminology, and I'm hoping to get rid of it some day. Maybe we should use the terminology "Account Menu" in our file naming and code so we don't introduce "MyResearch" into even more places.

File: config/vufind/MyResearchMenu.ini
Comment (xmorave2) [reply config/vufind/MyResearchMenu.ini]: Definitely. The first sentence is a leftover from time before I wrote the documentation.

File: config/vufind/MyResearchMenu.ini
Comment (xmorave2) [reply config/vufind/MyResearchMenu.ini]: Yes, that's how it should work. I'll update the documentation.

File: themes/bootstrap3/templates/myresearch/menu.phtml
Comment (demiankatz) [reply themes/bootstrap3/templates/myresearch/menu.phtml]: I see you haven't replied to this comment yet, but please let me know if you'd like me to prototype an example of what I'm talking about. I might be able to do this next week if you don't have time to look at it sooner. :-)

Comment (xmorave2) [reply themes/bootstrap3/templates/myresearch/menu.phtml]: I am working on it.

Comment (xmorave2) [reply themes/bootstrap3/templates/myresearch/menu.phtml]: I tried to make a helper, is this something you had in mind?

**Comment by xmorave2** [comment]
@ThoWagen Interetsting idea. This PR is very specific, but could be used to add some links to menu only when they would have a route in vufind, not for external ones. But could be relatively easily be modified to accept external links (and configurable ligthbox) too. If you have some use case for it, I can do it. But I was not planning to make it more general tool for adding links through the entire UI (although I agree it could be useful)

File: themes/bootstrap3/templates/myresearch/menu.phtml
Comment (demiankatz) [reply themes/bootstrap3/templates/myresearch/menu.phtml]: Thanks, @xmorave2, this looks like exactly what I had in mind. I'll review more closely and test a little later -- nearly time for the PMC meeting now. :-)

**Comment by ThoWagen** [comment]
@xmorave2 I think this is enough for the current PR. I just wanted to make sure that we don't do the same work twice :)

But I have one suggestion. Maybe it is better to use YAML instead of INI for the config. In my opinion this is the better choice for more complex configurations and that's how I planned to do it. I could also change that in a followup PR if that's wanted. But if you or @demiankatz prefer INI configs I don't mind that either.

**Comment by demiankatz** [comment]
@ThoWagen, I don't have a strong feeling about whether .ini or YAML is better for this case -- happy to go with whatever @xmorave2 prefers. I think both formats have trade-offs. As you say, YAML is better than .ini for describing complex configurations, especially if deep nesting of settings is required... but YAML is also a bit less user-friendly and more fragile than .ini in my experience (particularly when things break because of tiny whitespace issues), so I try to avoid it when it's not likely to be necessary.

**Comment by demiankatz** [comment]
@xmorave2, I'll wait on the final decision about configuration format before I do my most thorough review, but in the meantime I ran the full test suite (which passed) and made some minor adjustments to the comments in the .ini file to make them more complete based on what I saw elsewhere in the code.

One other idea for your consideration: might it be worth adding a `render(string $activeOption)` method on the helper, so we can replace all the:

```
<?=$this->context($this)->renderInContext('myresearch/menu.phtml', ['active' => 'librarycards'])?>
```

type calls with:

```
<?=$this->accountMenu()->render('librarycards')?>
```

? This might help raise awareness of the helper's existence, as well as make the code tidier and more consistent. If you prefer not to do this, though, I don't insist; it's just an idea!

**Comment by xmorave2** [comment]
I have moved the configuration to yaml and slightly changed its structure. I also added the render method to the helper

File: module/VuFind/src/VuFind/View/Helper/Root/AccountMenu.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/View/Helper/Root/AccountMenu.php]: You could simplify this condition to:

`!isset($item['checkMethod']) || $this->{$item['checkMethod']}()`

If you check the unset condition first, then you can just assume the value is set after the `||`.

File: themes/bootstrap3/templates/myresearch/menu.phtml
Comment (demiankatz) [inline themes/bootstrap3/templates/myresearch/menu.phtml]: Is this blank line needed? We don't usually have whitespace between the initial PHP block and subsequent markup.

File: module/VuFind/src/VuFind/View/Helper/Root/AccountMenu.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/View/Helper/Root/AccountMenu.php]: Not strictly necessary, but maybe it's worth adding a getUser support method to wrap around the isLoggedIn calls for greater clarity.

File: module/VuFind/src/VuFind/View/Helper/Root/AccountMenu.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/View/Helper/Root/AccountMenu.php]: You could replace `['patron' => $patron]` with `compact('patron')` as a trivial simplification.

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks, @xmorave2 -- this looks really good, and I've double-checked that the refactoring looks correct and the default settings are consistent with the default configuration file.

I haven't done hands-on testing yet -- first I had a few small questions and suggestions. You are free to ignore most of these if you don't think they are helpful, but at least one or two of the suggested optimizations might be worth the effort.

Once you've made whatever final changes you feel are worthwhile, and if @ThoWagen has no further suggestions, I'll do a hands-on test and merge if all goes well.

File: module/VuFind/src/VuFind/View/Helper/Root/AccountMenu.php
Comment (xmorave2) [reply module/VuFind/src/VuFind/View/Helper/Root/AccountMenu.php]: done

File: themes/bootstrap3/templates/myresearch/menu.phtml
Comment (xmorave2) [reply themes/bootstrap3/templates/myresearch/menu.phtml]: blank line removed

File: module/VuFind/src/VuFind/View/Helper/Root/AccountMenu.php
Comment (xmorave2) [reply module/VuFind/src/VuFind/View/Helper/Root/AccountMenu.php]: done

File: module/VuFind/src/VuFind/View/Helper/Root/AccountMenu.php
Comment (xmorave2) [reply module/VuFind/src/VuFind/View/Helper/Root/AccountMenu.php]: done

**Comment by xmorave2** [comment]
@demiankatz Thanks for the review, I've just incorporated your suggestion, so this should be ready for another round

**Comment by ThoWagen** [comment]
I haven't done any testing, but this looks good to me and I have no further suggestions. This should be a good starting point for me to implement additional features. Thanks!

**Review by demiankatz — APPROVED** [review]
Looks good to me -- thanks, @xmorave2!

**Comment by dltj** [comment]
Piping in to add a _THANK YOU_ to @xmorave2 for developing this. A nicely elegant way of customizing the Account Menu in configuration rather than code.

**Comment by demiankatz** [comment]
I'm glad you're finding this useful, @dltj. You might also be interested in checking out #4006, which extends things a little further. It's still in review/discussion, so please feel free to weigh in if you have opinions!

**Pull request merged**
